### PR TITLE
paywall is available as module

### PR DESCRIPTION
**Pull request by julien51**
# Description

Created a module which exports the `isUnlocked` method which can be used in different contexts
This is alpha release... as there are things to imporve!
Feel free to chime in!

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
Here is a script

```
const paywall = require('@unlock-protocol/paywall')

async function run() {
  const userAccountAddress = '0x62139339a2a3966c9660cab942e27f22b549f306'
  const paywallConfig = {
    locks: {
      '0xd3fCD4c1fA2952CF3B1a3f531C95b4b89F0974E6': {},
    },
  }
  const readOnlyProvider =
    'https://eth-mainnet.alchemyapi.io/jsonrpc/6idtzGwDtRbzil3s6QbYHr2Q_WBfn100'
  const locksmithUri = 'https://locksmith.unlock-protocol.com/'
  const unlocked = await paywall.default.isUnlocked(
    userAccountAddress,
    paywallConfig,
    {
      readOnlyProvider,
      locksmithUri,
    }
  )
  console.log({
    unlocked,
  })
}

run()

```

**Comment by julien51** [comment]
i do not like the `paywall.default.isUnlocked` but not sure how to avoid it :/

File: paywall/src/utils/isUnlocked.ts [hunk 1]
```diff
@@ -1,11 +1,5 @@
 import { keyExpirationTimestampFor } from './keyExpirationTimestampFor'
 import { optimisticUnlocking } from './optimisticUnlocking'
-
-declare let __ENVIRONMENT_VARIABLES__: {
```
Comment (julien51) [inline paywall/src/utils/isUnlocked.ts]: This cannot work in a module... so we are now passing these values to the function

File: paywall/src/paywall-script/module.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { isUnlocked } from '../utils/isUnlocked'
+
+export default {
```
Comment (julien51) [inline paywall/src/paywall-script/module.js]: Eventually, I think we could export the `Paywall` object too, which then front end applications could use directly!

File: paywall/src/paywall-script/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+          exclude: /node_modules/,
+        },
+      ],
+    },
+    resolve: {
+      extensions: ['.ts', '.js'],
+    },
+    plugins: [
+      new webpack.DefinePlugin({
+        fetch: "require('node-fetch')",
```
Comment (julien51) [inline paywall/src/paywall-script/webpack.config.js]: This feels *very* dirty but basically we replace any mention of `fetch` with `require('node-fetch')` at build time.

File: paywall/src/paywall-script/module.js (hunk 2 above)
Comment (cnasc) [inline paywall/src/paywall-script/module.js]: I believe if you don't do a default export you'll be able to reference it as 

```javascript
const paywall = require('@unlock-protocol/paywall')
paywall.isUnlocked // ...
```

Without having the default in between

Comment (julien51) [reply paywall/src/paywall-script/module.js]: but then how do I `export`?

Comment (cnasc) [reply paywall/src/paywall-script/module.js]: You just `export` without the default keyword

File: paywall/src/utils/isUnlocked.ts [hunk 4]
```diff
@@ -14,10 +8,9 @@ declare let __ENVIRONMENT_VARIABLES__: {
  */
 export const isUnlocked = async (
   userAccountAddress: string,
-  paywallConfig: any
+  paywallConfig: any,
```
Comment (cnasc) [inline paywall/src/utils/isUnlocked.ts]: https://github.com/unlock-protocol/unlock/blob/c65641fcd38078409b1b828645a924698482e335/paywall/src/unlockTypes.ts#L85

File: paywall/src/utils/isUnlocked.ts [hunk 5]
```diff
@@ -14,10 +8,9 @@ declare let __ENVIRONMENT_VARIABLES__: {
...
+  { readOnlyProvider, locksmithUri }: any
```
Comment (cnasc) [inline paywall/src/utils/isUnlocked.ts]: I think these should both be strings?

File: paywall/src/paywall-script/module.js (hunk 2 above)
Comment (julien51) [reply paywall/src/paywall-script/module.js]: oh! of course! let me try!

**Comment by cnasc** [comment]
LG overall with a few comments. I think it would be nice if we could also generate a `.d.ts` file to include in the module, would be a big benefit to any TS users on the consumption side.

File: paywall/src/paywall-script/module.js (hunk 2 above)
Comment (julien51) [reply paywall/src/paywall-script/module.js]: worked :)

File: paywall/src/paywall-script/module.d.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+interface PaywallModuleConfig {
+  readOnlyProvider: string
+  locksmithUri: string
+}
+
+interface PaywallModule {
```
Comment (cnasc) [inline paywall/src/paywall-script/module.d.ts]: I think you might want
```typescript
export function  isUnlocked(
    userAccountAddress: string,
    paywallConfig: string,
    config: PaywallModuleConfig
  ): Promise<boolean>
```

**Pull request merged**